### PR TITLE
Release/snowplow web/0.16.0

### PR DESCRIPTION
**Pull request by rlh1994**
## Description & motivation

**Note this will fail until utils is released and live**. Bringing every and anyone in to review to catch anything we can.

Release version 0.16.0 of web which contains:
- Migration to new base macro, enabling custom session and users identifiers (overwrite existing field), and custom fields from sde/entities in redshift/postgres
- New passthrough fields for core derived tables
- Increased flexibility of user mapping, and ability to run it on page views
- Change license to SCL

## Checklist
- [x] I have verified that these changes work locally
- [x] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)
- [x] I have raised a [documentation](https://github.com/snowplow/documentation) PR if applicable (Link [here](https://github.com/snowplow/documentation/pull/445) if required)
- [x] Is your change a breaking change?

## Release Only Checklist
- [x] I have updated the version number in all relevant places
- [x] I have updated the CHANGELOG.md

### CODE REVIEW

File: models/base/scratch/default/snowplow_web_base_events_this_run.sql [hunk 1]
```diff
@@ -6,313 +13,70 @@
...
-left join yauaa_context_dedupe as ya
-  on e.event_id = ya.yauaa_root_id and e.collector_tstamp = ya.yauaa_root_tstamp
-{%- endif %}
+{% set base_query_cols = get_column_schema_from_query( 'select * from (' + base_events_query +') a') %}
 
-where e.event_id_dedupe_index = 1
+select
+  {% for col in base_query_cols | map(attribute='name') | list -%}
+    {% if col == var("snowplow__page_view_context")~'_id' -%}
+      a.{{var("snowplow__page_view_context")}}_id as page_view_id
```
Comment (agnessnowplow) [inline models/base/scratch/default/snowplow_web_base_events_this_run.sql:67]: This doesn't do anything, the page_view_context is prefixed with 'page_view' so the column exists regardless.

Comment (rlh1994) [reply models/base/scratch/default/snowplow_web_base_events_this_run.sql:67]: Great point, I'll remove it!

File: CHANGELOG [hunk 2]
```diff
@@ -1,3 +1,36 @@
...
+- Renamed columns in quarantined sessions and sessions lifecycle manifest table (please run the ALTER TABLE statements in the migration guide or do a full-refresh)
+- Seeds have been renamed to be prefixed with `snowplow_web_*` to align with the rest of the package, and now you can use your own seed files or tables in their place (see the package config)
+
+## Features
+- Migrate base models to the new `base` macros for flexibility and consistency
+- Add ability to add custom entity/self-describing event columns to the `base_events_this_run` table for Redshift/Postgres
+- Add ability to use a custom session identifier and user identifier
+- Add new fields `original_domain_sessionid` and `original_domain_userid` to derived tables (where applicable) 
+- Add ability to pass fields through to derived page views, sessions, and user tables
+- Add option to choose the logged `user_id` field that session stitching uses
```
Comment (agnessnowplow) [inline CHANGELOG:24]: The page view stitching is missing here.

File: README.md
Comment (jpiekos) [inline README.md:1]: I believe this should be under our new Snowplow Academic License (SPAL) vs SCL.

Comment (jpiekos) [reply README.md:1]: After speaking with Mike, he has some strong reasons to keep it SCL and will follow up on the component licensing Notion page.

File: models/base/manifest/snowplow_web_base_sessions_lifecycle_manifest.sql [hunk 3]
```diff
@@ -21,82 +28,22 @@
...
-  sl.domain_userid,
-  sl.start_tstamp,
-  least({{ snowplow_utils.timestamp_add('day', var("snowplow__max_session_days", 3), 'sl.start_tstamp') }}, sl.end_tstamp) as end_tstamp -- limit session length to max_session_days
-  {% if target.type in ['databricks', 'spark'] -%}
-  , DATE(start_tstamp) as start_tstamp_date
-  {%- endif %}
-
-from session_lifecycle sl
+{% set sessions_lifecycle_manifest_query = snowplow_utils.base_create_snowplow_sessions_lifecycle_manifest(
+    session_identifiers=var('snowplow__session_identifiers', [{"table" : "events", "field" : "domain_sessionid"}]),
```
Comment (emielver) [inline models/base/manifest/snowplow_web_base_sessions_lifecycle_manifest.sql:32]:
```suggestion
    session_identifiers=var('snowplow__session_identifiers', [{"schema" : "atomic", "field" : "domain_sessionid"}]),
```

File: models/base/scratch/bigquery/snowplow_web_base_events_this_run.sql [hunk 4]
```diff
@@ -1,37 +1,39 @@
...
-
-from (
-
-  select
-    a.contexts_com_snowplowanalytics_snowplow_web_page_1_0_0[safe_offset(0)].id as page_view_id,
-    b.domain_userid, -- take domain_userid from manifest. This ensures only 1 domain_userid per session.
-    a.* except(contexts_com_snowplowanalytics_snowplow_web_page_1_0_0, domain_userid)
+{% set base_events_query = snowplow_utils.base_create_snowplow_events_this_run(
+    sessions_this_run_table='snowplow_web_base_sessions_this_run',
+    session_identifiers=var('snowplow__session_identifiers', [{"table" : "events", "field" : "domain_sessionid"}]),
```
Comment (emielver) [inline models/base/scratch/bigquery/snowplow_web_base_events_this_run.sql:16]:
```suggestion
    session_identifiers=var('snowplow__session_identifiers', [{"schema" : "atomic", "field" : "domain_sessionid"}]),
```

File: models/base/scratch/databricks/snowplow_web_base_events_this_run.sql [hunk 5]
```diff
@@ -1,25 +1,40 @@
...
-    a.contexts_com_snowplowanalytics_snowplow_web_page_1[0].id as page_view_id,
-    b.domain_userid, -- take domain_userid from manifest. This ensures only 1 domain_userid per session.
-    a.* except(contexts_com_snowplowanalytics_snowplow_web_page_1, a.domain_userid)
 
-from {{ var('snowplow__events') }} as a
-inner join {{ ref('snowplow_web_base_sessions_this_run') }} as b
-on a.domain_sessionid = b.session_id
+{% set base_events_query = snowplow_utils.base_create_snowplow_events_this_run(
+    sessions_this_run_table='snowplow_web_base_sessions_this_run',
+    session_identifiers=var('snowplow__session_identifiers', [{"table" : "events", "field" : "domain_sessionid"}]),
```
Comment (emielver) [inline models/base/scratch/databricks/snowplow_web_base_events_this_run.sql:17]:
```suggestion
    session_identifiers=var('snowplow__session_identifiers', [{"schema" : "atomic", "field" : "domain_sessionid"}]),
```

File: models/base/scratch/default/snowplow_web_base_events_this_run.sql [hunk 6]
```diff
@@ -6,313 +13,68 @@
...
-  {{snowplow_web.get_iab_context_fields('iab')}},
+{% if var('snowplow__enable_cwv', false) -%}
+  {% do contexts.append({'name': var('snowplow__cwv_context'), 'prefix': 'cwv', 'single_entity': True}) %}
+{% endif -%}
 
-  -- ua parser enrichment fields
-  {{snowplow_web.get_ua_context_fields('ua')}},
+{% set base_events_query = snowplow_utils.base_create_snowplow_events_this_run(
+    sessions_this_run_table='snowplow_web_base_sessions_this_run',
+    session_identifiers=var('snowplow__session_identifiers', [{"table" : "events", "field" : "domain_sessionid"}]),
```
Comment (emielver) [inline models/base/scratch/default/snowplow_web_base_events_this_run.sql:45]:
```suggestion
    session_identifiers=var('snowplow__session_identifiers', [{"schema" : "atomic", "field" : "domain_sessionid"}]),
```

**Review by agnessnowplow — APPROVED** [review]
LGTM

**Review by emielver — COMMENTED** [review]
Some small comments but good otherwise!

File: models/base/scratch/snowflake/snowplow_web_base_events_this_run.sql [hunk 7]
```diff
@@ -1,27 +1,40 @@
...
     sql_header=snowplow_utils.set_query_tag(var('snowplow__query_tag', 'snowplow_dbt'))
   )
 }}
 
-{%- set lower_limit, upper_limit = snowplow_utils.return_limits_from_model(ref('snowplow_web_base_sessions_this_run'),
-                                                                          'start_tstamp',
-                                                                          'end_tstamp') %}
+{% set base_events_query = snowplow_utils.base_create_snowplow_events_this_run(
+    sessions_this_run_table='snowplow_web_base_sessions_this_run',
+    session_identifiers=var('snowplow__session_identifiers', [{"table" : "events", "field" : "domain_sessionid"}]),
```
[the remark this answers is not in the document]
Comment (rlh1994) [reply models/base/scratch/snowflake/snowplow_web_base_events_this_run.sql:17]: Good catch, we should never hit the default this way but good to fix!

**Pull request merged**